src/data/set/basic.lean

```diff
@@ -1735,39 +1740,42 @@ ext $ by simp [set.prod]
 @[simp] theorem empty_prod : set.prod ∅ t = (∅ : set (α × β)) :=
 ext $ by simp [set.prod]
 
+theorem prod_inter_prod : set.prod s₁ t₁ ∩ set.prod s₂ t₂ = set.prod (s₁ ∩ s₂) (t₁ ∩ t₂) :=
+subset.antisymm
+  (assume ⟨a, b⟩ ⟨⟨ha₁, hb₁⟩, ⟨ha₂, hb₂⟩⟩, ⟨⟨ha₁, ha₂⟩, ⟨hb₁, hb₂⟩⟩)
+  (subset_inter
+    (prod_mono (inter_subset_left _ _) (inter_subset_left _ _))
+    (prod_mono (inter_subset_right _ _) (inter_subset_right _ _)))
+
 theorem insert_prod {a : α} {s : set α} {t : set β} :
   set.prod (insert a s) t = (prod.mk a '' t) ∪ set.prod s t :=
-ext begin simp [set.prod, image, iff_def, or_imp_distrib] {contextual := tt}; cc end
+begin ext ⟨x, y⟩, simp [set.prod, image, iff_def, or_imp_distrib, imp.swap] {contextual := tt} end
 
 theorem prod_insert {b : β} {s : set α} {t : set β} :
   set.prod s (insert b t) = ((λa, (a, b)) '' s) ∪ set.prod s t :=
-ext begin simp [set.prod, image, iff_def, or_imp_distrib] {contextual := tt}; cc end
+begin ext ⟨x, y⟩, simp [set.prod, image, iff_def, or_imp_distrib, imp.swap] {contextual := tt} end
 
 theorem prod_preimage_eq {f : γ → α} {g : δ → β} :
   set.prod (preimage f s) (preimage g t) = preimage (λp, (f p.1, g p.2)) (set.prod s t) := rfl
 
-theorem prod_mono {s₁ s₂ : set α} {t₁ t₂ : set β} (hs : s₁ ⊆ s₂) (ht : t₁ ⊆ t₂) :
-  set.prod s₁ t₁ ⊆ set.prod s₂ t₂ :=
-assume x ⟨h₁, h₂⟩, ⟨hs h₁, ht h₂⟩
+@[simp] lemma mk_preimage_prod_left {y : β} (h : y ∈ t) : (λ x, (x, y)) ⁻¹' s.prod t = s :=
+by { ext x, simp [h] }
 
-theorem prod_inter_prod : set.prod s₁ t₁ ∩ set.prod s₂ t₂ = set.prod (s₁ ∩ s₂) (t₁ ∩ t₂) :=
-subset.antisymm
-  (assume ⟨a, b⟩ ⟨⟨ha₁, hb₁⟩, ⟨ha₂, hb₂⟩⟩, ⟨⟨ha₁, ha₂⟩, ⟨hb₁, hb₂⟩⟩)
-  (subset_inter
-    (prod_mono (inter_subset_left _ _) (inter_subset_left _ _))
-    (prod_mono (inter_subset_right _ _) (inter_subset_right _ _)))
+@[simp] lemma mk_preimage_prod_right {x : α} (h : x ∈ s) : prod.mk x ⁻¹' s.prod t = t :=
+by { ext y, simp [h] }
 
 theorem image_swap_prod : (λp:β×α, (p.2, p.1)) '' set.prod t s = set.prod s t :=
-ext $ assume ⟨a, b⟩, by simp [mem_image_eq, set.prod, and_comm]; exact
-⟨ assume ⟨b', a', ⟨h_a, h_b⟩, h⟩, by subst a'; subst b'; assumption,
-  assume h, ⟨b, a, ⟨rfl, rfl⟩, h⟩⟩
+ext $ λ ⟨a, b⟩, by simp [mem_image_eq, set.prod, and_comm]; exact
+⟨ λ ⟨b', a', ⟨h_a, h_b⟩, h⟩, by subst a'; subst b'; assumption,
+  λ h, ⟨b, a, ⟨rfl, rfl⟩, h⟩⟩
```

Comment:
If you move this below the next lemma, then the proof will become trivial. In particular, this is a way to get rid of a non-finalizing `simp`.

Reply:
I didn't even notice the nonterminal simp. Good catch!